File: 프로그래머스/lv3/92343. 양과 늑대/양과 늑대.java
import java.io.*;
import java.util.*;

class Solution {
    List<List<Integer>> graph = new ArrayList<>();
    int answer = 0;
    void choose(int wolf, int sheep, int curNode, List<Integer> nextNodes, int[] info){
        if(info[curNode] == 0) sheep++;
        else wolf++;
        answer = Math.max(answer, sheep);
        if(wolf >=sheep) {
            // answer = Math.max(answer, sheep);
            return;
        }
        
        for(int i=0;i<nextNodes.size();i++){
            int nextNode = nextNodes.get(i);
            List<Integer> stackNodes = new ArrayList<>(nextNodes);
            stackNodes.remove((Integer)nextNode);
            stackNodes.addAll(graph.get(nextNode));
            choose(wolf, sheep, nextNode, stackNodes, info);
        }
        
    }
    public int solution(int[] info, int[][] edges) {
        
        for(int i=0;i<info.length;i++){
            graph.add(new ArrayList<>());
        }
            
        for(int[] e:edges){
            graph.get(e[0]).add(e[1]);
        }
        List<Integer> nextNodes = new ArrayList<>();
        nextNodes.addAll(graph.get(0));
        choose(0, 0, 0, nextNodes, info);
        
        return answer;
    }
}

File: 프로그래머스/lv3/92343. 양과 늑대/양과 늑대.py
max_sheep = 0

def dfs(current_node, sheep, wolf, info, tree, visited):
    global max_sheep
    if sheep <= wolf:
        return
    max_sheep = max(max_sheep, sheep)
    
    for next_node in tree.get(current_node, []):
        if visited[next_node]:
            continue
        visited[next_node] = True
        next_sheep = sheep + (info[next_node] == 0)
        next_wolf = wolf + (info[next_node] == 1)
        dfs(next_node, next_sheep, next_wolf, info, tree, visited)
        # 백트래킹: visited를 원상 복구합니다.
        visited[next_node] = False

def solution(info, edges):
    global max_sheep
    max_sheep = 0
    
    # 각 노드의 자식 노드 리스트를 딕셔너리로 만듭니다.
    tree = {}
    for parent, child in edges:
        if parent not in tree:
            tree[parent] = []
        tree[parent].append(child)
    
    visited = [False] * len(info)
    visited[0] = True
    
    dfs(0, 1, 0, info, tree, visited)
    return max_sheep